core/shared/src/main/scala/zio/Cause.scala

```diff
@@ -822,28 +822,64 @@ object Cause extends Serializable {
     final case class Filter[E](p: Cause[E] => Boolean) extends Folder[Any, E, Cause[E]] {
       def empty(context: Any): Cause[E] = Cause.empty
 
-      def failCase(context: Any, error: E, stackTrace: StackTrace): Cause[E] = Cause.Fail(error, stackTrace)
+      def failCase(context: Any, error: E, stackTrace: StackTrace): Cause[E] = {
+        val c = Cause.Fail(error, stackTrace)
+        if (p(c))
+          c
+        else
+          Cause.empty
+      }
 
-      def dieCase(context: Any, t: Throwable, stackTrace: StackTrace): Cause[E] = Cause.Die(t, stackTrace)
+      def dieCase(context: Any, t: Throwable, stackTrace: StackTrace): Cause[E] = {
+        val c = Cause.Die(t, stackTrace)
+        if (p(c))
+          c
+        else
+          Cause.empty
+      }
 
-      def interruptCase(context: Any, fiberId: FiberId, stackTrace: StackTrace): Cause[E] =
-        Cause.Interrupt(fiberId, stackTrace)
+      def interruptCase(context: Any, fiberId: FiberId, stackTrace: StackTrace): Cause[E] = {
+        val c = Cause.Interrupt(fiberId, stackTrace)
+        if (p(c))
+          c
+        else
+          Cause.empty
+      }
 
       def bothCase(context: Any, left: Cause[E], right: Cause[E]): Cause[E] =
-        if (p(left)) {
-          if (p(right)) Cause.Both(left, right)
-          else left
-        } else if (p(right)) right
-        else Cause.empty
+        (left, right) match {
```

Comment:
Let's avoid the intermediate tuple creation here and in the methods below and use simple `if (left eq Cause.Empty) { ... } else if (right eq Cause.Empty) { ... } else {  ...  }` statements

Reply:
done